core/trino-main/src/test/java/io/trino/sql/query/TestArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,158 @@
+                                                                ImmutableList.of("a"),
+                                                                ImmutableList.of(
+                                                                        ImmutableList.of(new LongLiteral("1")),
+                                                                        ImmutableList.of(new LongLiteral("1")))))))))));
+    }
+
+    @Test
+    public void multipleArrayDistinctArrayAggRuleFiresTwice()
+    {
+        assertions.assertQueryAndPlan(
+                "SELECT ARRAY_DISTINCT(ARRAY_AGG(a)), ARRAY_DISTINCT(ARRAY_AGG(DISTINCT b)) FROM (VALUES (1, 2), (1, 2)) t(a, b)",
+                "VALUES (ARRAY [1], ARRAY [2])",
+                anyTree(aggregation(
+                        ImmutableMap.of("array_agg", functionCall("array_agg", ImmutableList.of("a")),
+                                "array_agg2", functionCall("array_agg", ImmutableList.of("a"))),
+                        anyTree(project(ImmutableMap.of("hash", expression("combine_hash(bigint '0', coalesce(\"$operator$hash_code\"(a), 0))"),
+                                        "hash2", expression("combine_hash(bigint '0', coalesce(\"$operator$hash_code\"(b), 0))")),
+                                values(
+                                        ImmutableList.of("a", "b"),
+                                        ImmutableList.of(
+                                                ImmutableList.of(new LongLiteral("1"), new LongLiteral("2")),
+                                                ImmutableList.of(new LongLiteral("1"), new LongLiteral("2")))))))));
+    }
+
+    @Test
+    public void multipleArrayDistinctArrayAggDifferentOrderRuleFiresTwice()
+    {
+        assertions.assertQueryAndPlan(
+                "SELECT ARRAY_DISTINCT(ARRAY_AGG(b)), ARRAY_DISTINCT(ARRAY_AGG(a)) FROM (VALUES (1, 2), (1, 2)) t(a, b)",
+                "VALUES (ARRAY [2], ARRAY [1])",
+                anyTree(aggregation(
+                        ImmutableMap.of("array_agg", functionCall("array_agg", ImmutableList.of("a")),
+                                "array_agg2", functionCall("array_agg", ImmutableList.of("a"))),
+                        anyTree(project(ImmutableMap.of("hash", expression("combine_hash(bigint '0', coalesce(\"$operator$hash_code\"(a), 0))"),
+                                        "hash2", expression("combine_hash(bigint '0', coalesce(\"$operator$hash_code\"(b), 0))")),
+                                values(
+                                        ImmutableList.of("a", "b"),
+                                        ImmutableList.of(
+                                                ImmutableList.of(new LongLiteral("1"), new LongLiteral("2")),
+                                                ImmutableList.of(new LongLiteral("1"), new LongLiteral("2")))))))));
+    }
+
+    @Test
+    public void singleArrayDistinctArrayAggMultipleAssignmentsRuleDoesNotFire()
```

Comment:
The "does not fire" test cases should go to a unit test for the rule. There you can use the dedicated method `doesNotFire()`, which lets you not check the plan in case when the rule didn't fire.

I think it would be also good to move plan testing to the unit test -- it lets you check the immediate result of the application of the rule, not the final optimized plan.

Regarding the test cases in this class, please keep them, but only test the query results -- consider it a correctness check, and not a test for the mechanics of the rule.